### ISSUE TITLE
[Feature request] Custom loss for PGD

### ISSUE THREAD

**Issue by CNOCycle** [tensorflow/cleverhans#1116]
Hi authors,

I'm developing a robust model against adversarial attacks.

Since our model is not trained by the normal PGD adversarial training, we should carefully verify that gradient masking or some effects do not occur in our model.

In some cases, our model can defense PGD attack very well but it would be defeated by PGD with CW loss which proposed by Madry.

Currently, I manually modify the loss function in FGM attack which is called by PGD attack when evalutating robust accuracy against PGD with CW loss.

https://github.com/tensorflow/cleverhans/blob/51cb6d660b9d333ee637baf595205c7f5d1f1d66/cleverhans/attacks/fast_gradient_method.py#L184-L185

For others competitive package, it allows user pass custom loss in PGD attack.

Similarly, I hope that cleverhans allows developers pass the custom loss function as parameter to PGD or FGM attack.

If developers do not specify the loss function, it should be the normal cross-entropy loss.

Is this thing possible? If not, please give me some suggestions.

Any help will be appreciated.

Thanks.

**Comment by iamgroot42** [comment]
Created a PR for this : https://github.com/tensorflow/cleverhans/pull/1120
Hopefully it will be merged if all tests pass :)

**Comment by CNOCycle** [comment]
Thank you for helping me.

I can run PGD with CW loss now from your PR.

Also, following is a minimal test case whose result is as expected.

```python
#%%
import time
import numpy as np
import tensorflow as tf

import cleverhans.attacks as c_attacks
from cleverhans.utils_keras import KerasModelWrapper

#%%
def build_model():

    # declare variables
    model_layers = [ tf.keras.layers.Conv2D(16, (3, 3), padding="same", activation="relu", kernel_initializer=tf.keras.initializers.he_normal()),
                     tf.keras.layers.Conv2D(32, (5, 5), padding="same", activation="relu", kernel_initializer=tf.keras.initializers.he_normal()),
                     tf.keras.layers.MaxPooling2D(pool_size=(2, 2), name="clf/p1"),
                     tf.keras.layers.Conv2D(32, (3, 3), padding="same", activation="relu", kernel_initializer=tf.keras.initializers.he_normal()),
                     tf.keras.layers.Conv2D(64, (5, 5), padding="same", activation="relu", kernel_initializer=tf.keras.initializers.he_normal()),
                     tf.keras.layers.MaxPooling2D(pool_size=(2, 2), name="clf/p2"),
                     tf.keras.layers.Flatten(name="clf/f1"),
                     tf.keras.layers.Dense(256, activation="relu", kernel_initializer=tf.keras.initializers.he_normal()),
                     tf.keras.layers.Dense(10 , activation=None  , kernel_initializer=tf.keras.initializers.he_normal()),
                     tf.keras.layers.Activation('softmax', name="clf/a1")
                    ]

    # clf_model
    optimizer = tf.keras.optimizers.Nadam(lr=0.001, beta_1=0.9, beta_2=0.99, schedule_decay=0.004)
    clf_input = tf.keras.layers.Input(shape=(28,28,1 ), name="model/input")
    clf_out   = clf_input
    for ii in model_layers:
        clf_out = ii(clf_out)
    clf_model = tf.keras.models.Model(inputs=clf_input, outputs=clf_out, name='clf_model')
    clf_model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    clf_model.summary()

    return clf_model

def get_data():

    # get mnist from keras
    (trainX, trainY), (testX, testY) = tf.keras.datasets.mnist.load_data()
    trainX = trainX.astype(np.float32)
    testX  = testX.astype(np.float32)

    # ont-hot
    trainY = tf.keras.utils.to_categorical(trainY,10)
    testY  = tf.keras.utils.to_categorical(testY ,10)

    # get validation sets
    validX = trainX[55000:,:]
    validY = trainY[55000:,:]
    trainX = trainX[:55000,:]
    trainY = trainY[:55000,:]

    # expand dimesion
    trainX = np.expand_dims(trainX, axis=3)
    validX = np.expand_dims(validX, axis=3)
    testX  = np.expand_dims(testX , axis=3)

    # normalization
    img_max = 1
    img_min = 0
    trainX = trainX / 255.0 * (img_max - img_min) + img_min
    validX = validX / 255.0 * (img_max - img_min) + img_min
    testX  = testX  / 255.0 * (img_max - img_min) + img_min
    return trainX, trainY, validX, validY, testX, testY, img_min, img_max

#%%
if __name__ == '__main__':

    import os
    if os.environ.get("CUDA_VISIBLE_DEVICES") is None:
        os.environ["CUDA_VISIBLE_DEVICES"] = "0"

    # reset tf session
    tf.keras.backend.clear_session()
    gpu_options = tf.GPUOptions(allow_growth=True)
    sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options))
    tf.keras.backend.set_session(sess)

    # get data and create model
    trainX, trainY, validX, validY, testX, testY, img_min, img_max = get_data()
    model = build_model()

    # train model - warm up
    warmup = True
    if warmup:
        model.fit(trainX, trainY,
                validation_data=(validX, validY),
                batch_size=1000,
                epochs=3,
                shuffle=True,
                verbose=2)

    # create attacking method
    wrap = KerasModelWrapper(model)
    wrap.model.summary()
    pgd = c_attacks.ProjectedGradientDescent(wrap, sess=sess)

    # select a targetd image
    x = testX[0]
    y = testY[0]
    x_4d = np.expand_dims(x, axis = 0)
    w = model.predict(x_4d)[0]
    print('clean label:[{:d}/{:d}]'.format(np.argmax(y), np.argmax(w)) )

    # generate normal adversarial image
    adv_x_ce = pgd.generate_np(x_4d,
                              eps=0.3,
                              eps_iter=0.01,
                              ord = np.inf,
                              nb_iter=10,
                              rand_init=True,
                              clip_min=img_min,
                              clip_max=img_max)
    ce_label = model.predict(adv_x_ce)[0]
    print('ce label: {:d}'.format(np.argmax(ce_label) ) )

    # generate adversarial image with CW loss
    def cw_loss(labels, logits):
        label_mask = labels
        correct_logit = tf.reduce_sum(label_mask * logits, axis=1)
        wrong_logit = tf.reduce_max((1-label_mask) * logits - 1e4*label_mask, axis=1)
        loss = -tf.nn.relu(correct_logit - wrong_logit + 50)
        return loss

    x = testX[0]
    adv_x_cw = pgd.generate_np(x_4d,
                              eps=0.3,
                              eps_iter=0.01,
                              ord = np.inf,
                              nb_iter=50,
                              rand_init=True,
                              loss_fn=cw_loss,
                              clip_min=img_min,
                              clip_max=img_max)
    cw_label = model.predict(adv_x_cw)[0]
    print('cw label: {:d}'.format(np.argmax(cw_label) ) )
```
